### ISSUE TITLE
Logging enhancements

### ISSUE THREAD

**Issue by thehesiod** [aio-libs/aiohttp#1294]
## Long story short

We're trying to output our logs in JSON by passing a dictionary to the logger.  Right now in order to do this we need to override a lot of classes to get to the logger.

Ex:

``` python
class ReqAccessJSONLogger(aiohttp.helpers.AccessLogger):
    def log(self, message, environ, response, transport, time):
        args = [message, environ, response, transport, time]
        try:
            obj = {
                "message": self._log_format % self._format_line([message, environ, response, transport, time]),
                "extra": {
                    "remoteIP": self._format_a(args),
                    "elapsedS": self._format_Tf(args),
                    "request": self._format_r(args),
                    "size": self._format_b(args),
                    "referrer": self._format_i('Referrer', args),
                    "userAgent": self._format_i('User-Agent', args),
                }
            }

            self.logger.info(obj)
        except BaseException:
            self.logger.exception("Error in logging")

class ReqHandler(aiohttp.web.RequestHandler):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        # here we replace the access logger
        self.access_logger = ReqAccessJSONLogger(self.access_log, aiohttp.helpers.AccessLogger.LOG_FORMAT)


class ReqHdlrFactory(aiohttp.web.RequestHandlerFactory):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs, handler=ReqHandler)

aiohttp.web.Application(logger=logger, loop=loop, handler_factory=ReqHdlrFactory, middlewares=[_middleware_factory])
```
## Expected behaviour

I should be a lot easer to implement this functionality and shouldn't require us to override `aiohttp.web.RequestHandler.access_logger` as this attribute is undocumented.

One option would be to allow for a `handler` kwarg in `aiohttp.web.run_app` to be passed in its call to make_handler so we don't need to override the handler_factory.  Another better option would be to allow for something like 'access_log_class` to be passed eventually to server.ServerHttpProtocol to be used with a default of helpers.AccessLogger.

Another is it would be nice to formalize the AccessLogger._format methods, right now it's dangerous to use them as they can change and I don't want to duplicate the implementation.

**Comment by thehesiod** [comment]
if direction is given I can come up with a PR, thanks!

**Comment by asvetlov** [comment]
I think the right solution is constructing an `extra` dict first and passing it to along with logging message.
Logger handler may do everything with given `extra` structure.
It doesn't require overriding `access_log_class` or doing other weird things.

The only tricky part is: for sake of performance `extra` should contain attributes specified by logger format string only.

**Comment by thehesiod** [comment]
sorry I'm not following, which class should construct this `extra` dict and pass it to which other class?  Could you show a small snippet of how you see this working?  I can then work off that.  Thanks!

**Comment by asvetlov** [comment]
`AccessLogger.log` should look like:

```
    def log(self, message, environ, response, transport, time):
        try:
            extra = self._collect_extra(message, environ, response, transport, time)
            # we can keep .format_line but passing a dict makes a code much easier to read
            msg = self._format_msg(extra)  
            self.logger.info(msg, extra=extra)
        except Exception:
            self.logger.exception("Error in logging")
```

**Comment by thehesiod** [comment]
ok, is something like this what you meant: https://github.com/KeepSafe/aiohttp/pull/1303?

**Comment by asvetlov** [comment]
Fixed by #1303
